### PR TITLE
retry signing after current signing session fails

### PR DESCRIPTION
**Pull request by BatiGencho**

### CODE REVIEW

File: bin/test-suite/src/suite/consensus/frost/test_frost_e2e.rs [hunk 1]
```diff
@@ -324,3 +326,248 @@ pub async fn frost_e2e(suite: &ConsensusIntegrationTestSuite) -> Result<(), supe
...
+    tokio::time::sleep(Duration::from_secs(3)).await;
+
+    // now disconnect the peers of fed member 0
+    test_fed_members.get(&0).cloned().unwrap().send_test_signal(TestSignal::DisconnectAll());
+
+    // wait for a new epoch to start
+    tokio::time::sleep(Duration::from_secs(10)).await;
+
+    // now reconnect the peers of fed member 0
+    test_fed_members.get(&0).cloned().unwrap().send_test_signal(TestSignal::DisconnectAll());
```
Comment (arminsabouri) [inline bin/test-suite/src/suite/consensus/frost/test_frost_e2e.rs]: The comment says reconnect but the signal is disconnect

File: bin/test-suite/src/suite/consensus/frost/test_frost_e2e.rs [hunk 2]
```diff
@@ -324,3 +326,248 @@ pub async fn frost_e2e(suite: &ConsensusIntegrationTestSuite) -> Result<(), supe
...
+    let pegout_destination =
+        ethers::core::types::Bytes::from(btc_address.to_string().as_bytes().to_vec());
+    // use empty pegout data
+    let pegout_data = ethers::core::types::Bytes::new();
+    let pegout_amount = Amount::from_btc(0.5).unwrap();
+    let tx_receipt =
+        mint_contract.burn(pegout_destination, pegout_data, pegout_amount.to_wei()).await.unwrap();
+    it_info_print!("Pegout Tx Receipt: ", tx_receipt);
+
+    // ===================== FAILURE =====================
```
Comment (arminsabouri) [inline bin/test-suite/src/suite/consensus/frost/test_frost_e2e.rs]: Should this reconnect / disconnect be happening after L521?
Where we send the final eoa tx?

File: crates/consensus/authority/src/signing.rs [hunk 3]
```diff
@@ -500,13 +610,52 @@ where
         psbt: Vec<u8>,
     ) -> Result<(), Error> {
         let session_id = parse_signing_session_id(&signing_session_id)?;
-        self.insert_signing_state(session_id, SigningState::Initial);
-        info!(">>>>>>>>>>> [START NEW SIGNING SESSION]");
+
+        // the coordinator is always expected to be us in this case, i.e. None
+        let coordinator = self.get_coordinator().await?;
+        if coordinator.is_some() {
+            panic!(">>>>>>>>>>> A non-coordinator is trying to (re)initiate a signing process!");
```
Comment (arminsabouri) [inline crates/consensus/authority/src/signing.rs]: Should we be panicing here? What If this is just a timing error?

File: crates/consensus/authority/src/utils.rs [hunk 4]
```diff
@@ -495,12 +520,16 @@ pub(crate) enum GenerateSigningSesssionIdError {
 pub(crate) fn generate_signing_session_id(
     best_hash: &[u8],
     authority_pk: &secp256k1::PublicKey,
+    has_timestamp: bool,
 ) -> Result<SigningSessionId, GenerateSigningSesssionIdError> {
     let mut engine = sha256::HashEngine::default();
```
Comment (arminsabouri) [inline crates/consensus/authority/src/utils.rs]: I feel like this function ended up getting more complicated than it needs to for its intended usecase.
How do you feel about just hashing uuidv4()

File: crates/consensus/common/src/utils.rs [hunk 5]
```diff
@@ -295,6 +295,23 @@ pub fn is_inturn(authorities_len: u64, signer_index: u64) -> bool {
     (timestamp / authorities_len) % authorities_len == signer_index
 }
 
+/// Returns the inturn interval for a signer index
+pub fn get_in_turn_interval(authorities_len: u64, signer_index: u64) -> (u64, u64, u64, u64) {
```
Comment (arminsabouri) [inline crates/consensus/common/src/utils.rs]: Can we typedef `(u64, u64, u64, u64)` to somethign like `CoordinatorInterval`?

File: bin/test-suite/src/suite/consensus/frost/test_frost_e2e.rs (hunk 1 above)
Comment (BatiGencho) [reply bin/test-suite/src/suite/consensus/frost/test_frost_e2e.rs]: correct, sorry. I have fixed it now.

File: bin/test-suite/src/suite/consensus/frost/test_frost_e2e.rs (hunk 2 above)
Comment (BatiGencho) [reply bin/test-suite/src/suite/consensus/frost/test_frost_e2e.rs]: I have re-arranged this logic now.

File: crates/consensus/authority/src/signing.rs (hunk 3 above)
Comment (BatiGencho) [reply crates/consensus/authority/src/signing.rs]: i made an error out of it. It should not happen though

File: crates/consensus/authority/src/utils.rs (hunk 4 above)
Comment (BatiGencho) [reply crates/consensus/authority/src/utils.rs]: done. Much cleaner

File: crates/consensus/common/src/utils.rs (hunk 5 above)
Comment (BatiGencho) [reply crates/consensus/common/src/utils.rs]: done. Thx for the suggestion

File: crates/consensus/authority/src/signing.rs [hunk 6]
```diff
@@ -545,49 +695,99 @@ where
...
+                .unwrap_or_default();
+
+            // abort this previous session as it is outdated
+            if unix_timestamp() >= prev_validity_to {
+                self.abort_signing().await?;
+                self.update_signing_state(session_id, SigningState::Failed);
+                return Ok(());
+            }
+
+            // session exists and is valid, return if we are not in round 2
```
Comment (scottmillner) [inline crates/consensus/authority/src/signing.rs:745]: nit: should say if we are not in round 1

it confused me for a second lol

**Review by scottmillner — DISMISSED** [review]
as always 🔥 🔥 🔥 🔥 🔥 
awesome sauce!!

File: crates/net/network/src/frost/protocol.rs [hunk 7]
```diff
@@ -163,6 +163,7 @@ impl Stream for FrostProtoConnection {
             return Poll::Ready(Some(initial_ping.encoded()))
         }
         let peer_message_forwarder = this.peer_message_forwarder.clone();
+        //loop {
```
Comment (arminsabouri) [inline crates/net/network/src/frost/protocol.rs]: Nit: remove commented out code

File: crates/consensus/authority/src/signing.rs [hunk 8]
```diff
@@ -144,43 +165,93 @@ where
...
+    /// Checks if a signing session exists or not yet
+    pub(crate) fn signing_session_exists(&mut self, session_id: [u8; 32]) -> bool {
+        self.signing_sessions.contains_key(&session_id)
+    }
+
+    /// Inserts a new signing session
+    pub(crate) fn insert_new_signing_session(
         &mut self,
         session_id: [u8; 32],
+        coordinator_id: u64,
```
Comment (arminsabouri) [inline crates/consensus/authority/src/signing.rs]: nit: is this an index into the fed list? if so it should be `coordiantor_idnex`

File: crates/consensus/authority/src/signing.rs [hunk 9]
```diff
@@ -504,13 +614,53 @@ where
...
+        if coordinator.is_some() {
+            error!(">>>>>>>>>>> A non-coordinator is trying to (re)initiate a signing process!");
+            return Ok(());
+        }
+        let (start, end, time_passed, time_remaining) =
+            self.get_inturn_interval_for_coordinator(None);
+
+        // try to find the signing session in cache in case it was re-triggered by the same
+        // coordinator
+        match self.get_signing_session(session_id) {
```
Comment (arminsabouri) [inline crates/consensus/authority/src/signing.rs:629]: What actually causes this to be re-triggered. IIRC the block builder just waits 60s for the current signing session to pass?

**Review by arminsabouri — COMMENTED** [review]
Awesome work. Just had a few questions

File: crates/consensus/authority/src/signing.rs (hunk 9 above)
Comment (BatiGencho) [reply crates/consensus/authority/src/signing.rs:629]: check the logic here: https://github.com/botanix-labs/Macbeth/pull/168/files#diff-f459d49b88e4803ede826537533f5067794f1d39165d9791add7a6bb6b3ac777R1017-R1067

if the current coordinator is still in turn and it decides there is time, it would retry it.

File: crates/consensus/authority/src/signing.rs (hunk 8 above)
Comment (BatiGencho) [reply crates/consensus/authority/src/signing.rs]: of course, fixed

File: crates/net/network/src/frost/protocol.rs (hunk 7 above)
Comment (BatiGencho) [reply crates/net/network/src/frost/protocol.rs]: sure, done

File: .github/workflows/unit.yml [hunk 10]
```diff
@@ -20,52 +20,6 @@ concurrency:
...
-      - name: Install Protobuf Compiler
-        run: sudo apt-get update && sudo apt-get install -y protobuf-compiler
-      - uses: dtolnay/rust-toolchain@stable
-      - uses: taiki-e/install-action@nextest
-      - uses: Swatinem/rust-cache@v2
-        with:
-          cache-on-failure: true
-      - name: Run e2e tests
-        run: |
-          make start-test-suite
```
Comment (arminsabouri) [inline .github/workflows/unit.yml]: Did we mean to delete this? Is this being run somewhere else?

File: bin/test-suite/src/suite/consensus/frost/test_frost_e2e_edge_cases.rs [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+
+    // now disconnect the peers of fed member 0
+    test_fed_members.get(&0).cloned().unwrap().send_test_signal(TestSignal::DisconnectAll());
+
+    // wait for a new epoch to start
+    tokio::time::sleep(Duration::from_secs(10)).await;
+
+    // now reconnect the peers of fed member 0
+    test_fed_members.get(&0).cloned().unwrap().send_test_signal(TestSignal::ReconnectAll());
+    tokio::time::sleep(Duration::from_secs(3)).await;
```
Comment (arminsabouri) [inline bin/test-suite/src/suite/consensus/frost/test_frost_e2e_edge_cases.rs:240]: What are we testing here? That we can perform a signing round without a peer being connected?

Comment (arminsabouri) [reply bin/test-suite/src/suite/consensus/frost/test_frost_e2e_edge_cases.rs:240]: It would be helpful to have a comment above the test function describing what exact edge case we are testing

**Comment by arminsabouri** [comment]
Don't mean to be a commit Nazi. But the commits here could probably be squashed into one or maybe two 
- signing edge cases
- Fixed int pipeline

File: crates/botanix-lib/src/extra_data_header.rs [hunk 12]
```diff
@@ -368,11 +368,28 @@ mod tests {
         );
         let mut buf: Vec<u8> = vec![];
         header.encode_into_without_signature(&mut buf).unwrap();
-        println!("{:?}", buf);
-        // serialize the same header
-        let serialized =
-            ExtraDataHeader::deserialize(&mut buf.as_slice()).expect("Deserialization");
-        assert_eq!(serialized, header);
+        // Check version
```
Comment (arminsabouri) [inline crates/botanix-lib/src/extra_data_header.rs:371]: I simplified this test in a earlier commit b/c I don't think doing exact bit by bit matching really got us much.
I think testing if something deserializes and serializes to the same thing is suffecient.
Lmk if you disagree

File: .github/workflows/unit.yml (hunk 10 above)
Comment (arminsabouri) [reply .github/workflows/unit.yml]: I just saw this was moved to unit.yml
Closing

**Review by arminsabouri — APPROVED** [review]

File: bin/test-suite/src/suite/consensus/frost/test_frost_e2e_edge_cases.rs (hunk 11 above)
Comment (BatiGencho) [reply bin/test-suite/src/suite/consensus/frost/test_frost_e2e_edge_cases.rs:240]: Sure, I will add more comments in the next commit!

File: crates/botanix-lib/src/extra_data_header.rs (hunk 12 above)
Comment (BatiGencho) [reply crates/botanix-lib/src/extra_data_header.rs:371]: I totally agree, all good!
